slay/definitions.py

```diff
@@ -0,0 +1,259 @@
+    memory: str = "2Gi"
+    gpu: Optional[str] = None
+
+
+class Compute:
+    """Builder to create compute spec."""
+
+    _spec: ComputeSpec
+
+    def __init__(self) -> None:
+        self._spec = ComputeSpec()
+
+    def cpu(self, cpu: int) -> "Compute":
+        self._spec.cpu = str(cpu)
+        return self
+
+    def memory(self, memory: str) -> "Compute":
+        self._spec.memory = memory
+        return self
+
+    def gpu(self, kind: str, count: int = 1) -> "Compute":
+        self._spec.gpu = f"{kind}:{count}"
+        return self
+
+    def get_spec(self) -> ComputeSpec:
+        return self._spec.copy(deep=True)
+
+
+class AssetSpec(pydantic.BaseModel):
+    # TODO: this is not stable yet and might change or refer back to truss.
+    secrets: dict[str, str] = {}
+    cached: list[Any] = []
+
+
+class Assets:
+    """Builder to create asset spec."""
+
+    _spec: AssetSpec
+
+    def __init__(self) -> None:
+        self._spec = AssetSpec()
+
+    def secret(self, key: str) -> "Assets":
+        self._spec.secrets[key] = "***"
+        return self
+
+    def cached(self, value: list[Any]) -> "Assets":
+        self._spec.cached = value
+        return self
+
+    def get_spec(self) -> AssetSpec:
+        return self._spec.copy(deep=True)
+
+
+class Config(generics.GenericModel, Generic[UserConfigT]):
+    """Bundles config values needed to deploy a processor."""
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    name: Optional[str] = None
+    image: Image = Image()
+    compute: Compute = Compute()
+    assets: Assets = Assets()
+    user_config: UserConfigT = pydantic.Field(default=None)
+
+    def get_image_spec(self) -> ImageSpec:
+        return self.image.get_spec()
+
+    def get_compute_spec(self) -> ComputeSpec:
+        return self.compute.get_spec()
+
+    def get_asset_spec(self) -> AssetSpec:
+        return self.assets.get_spec()
+
+
+class Context(generics.GenericModel, Generic[UserConfigT]):
+    """Bundles config values needed to instantiate a processor in deployment."""
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    user_config: UserConfigT = pydantic.Field(default=None)
+    stub_cls_to_url: Mapping[str, str] = {}
+    # secrets: Optional[secrets_resolver.Secrets] = None
+    # TODO: above type results in `truss.server.shared.secrets_resolver.Secrets`
+    # due to the templating, at runtime the object passed will be from
+    # `shared.secrets_resolver` and give pydantic validation error.
+    secrets: Optional[Any] = None
+
+    def get_stub_url(self, stub_cls_name: str) -> str:
+        if stub_cls_name not in self.stub_cls_to_url:
+            raise MissingDependencyError(f"{stub_cls_name}")
+        return self.stub_cls_to_url[stub_cls_name]
+
+    def get_baseten_api_key(self) -> str:
+        if not self.secrets:
+            raise UsageError(f"Secrets not set in `{self.__class__.__name__}` object.")
+        if BASTEN_APY_SECRET_NAME not in self.secrets:
```

Comment:
does this mean that all workflow deployments must be `--trusted`? that's probably okay, but we have to make sure that we do that for users automatically

Reply:
Yes, with the current decision of using an API_KEY - it's already done automatically in the deploy command.